### ISSUE TITLE
Error in Fiber.interruptAll

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#1440]
Reported by @ghostdogpr 

```
- offers are suspended by back pressure
[info]     Fiber failed.
[info]     An unchecked error was produced.
[info]     java.lang.NullPointerException
[info]     	at zio.Fiber$.$anonfun$interruptAll$2(Fiber.scala:296)
[info]     	at zio.ZIO$ZipLeftFn.apply(ZIO.scala:2490)
[info]     	at zio.ZIO$ZipLeftFn.apply(ZIO.scala:2488)
[info]     	at zio.internal.FiberContext.evaluateNow(FiberContext.scala:270)
[info]     	at zio.internal.FiberContext.$anonfun$evaluateLater$1(FiberContext.scala:602)
[info]     	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[info]     	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[info]     	at java.lang.Thread.run(Thread.java:748)
[info]     
[info]     Fiber:12547 was supposed to continue to:
[info]       a future continuation at zio.ZIO.ensuring(ZIO.scala:285)
[info]       a future continuation at zio.ZIO.run(ZIO.scala:1007)
[info]       a future continuation at zio.ZIO.bracket_(ZIO.scala:131)
[info]       a future continuation at zio.ZIO.run(ZIO.scala:1007)
[info]       a future continuation at zio.ZIO.bracket_(ZIO.scala:131)
[info]       a future continuation at zio.ZIO.run(ZIO.scala:1007)
[info]       a future continuation at zio.ZManaged.use(ZManaged.scala:634)
[info]       a future continuation at zio.test.TestExecutor$.managed(TestExecutor.scala:25)
[info]       a future continuation at zio.test.Spec.foreachExec(Spec.scala:95)
[info]       a future continuation at zio.ZIOFunctions.foreachParN(ZIO.scala:1896)
[info]       a future continuation at zio.ZIO.forever(ZIO.scala:406)
[info]     
[info]     Fiber:12547 execution trace:
[info]       at zio.Fiber$.interruptAll(Fiber.scala:296)
[info]       at zio.ZIOFunctions.handleChildrenWith(ZIO.scala:2030)
[info]       at zio.internal.FiberContext.getFibers(FiberContext.scala:560)
[info]       at zio.ZIOFunctions.children(ZIO.scala:1595)
[info]       at zio.ZIO.ensuring(ZIO.scala:283)
[info]       at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:146)
[info]       at zio.ZQueueSpec$$anonfun$$lessinit$greater$1.new(ZQueueSpec.scala:55)
[info]       at zio.Ref$.get$extension(Ref.scala:39)
[info]       at zio.ZQueueSpec$$anonfun$$lessinit$greater$1.new(ZQueueSpec.scala:54)
[info]       at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:155)
[info]       at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:146)
[info]       at zio.ZIOFunctions.bracketExit(ZIO.scala:1563)
[info]       at zio.internal.FiberContext.evaluateNow(FiberContext.scala:473)
[info]       at zio.ZIO.bracket_(ZIO.scala:131)
[info]       at zio.ZIO.run(ZIO.scala:1007)
[info]       at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:146)
[info]       at zio.ZIO.repeatOrElse(ZIO.scala:930)
[info]       at zio.ZIO.repeatOrElseEither(ZIO.scala:946)
[info]       at zio.ZSchedule.check(ZSchedule.scala:211)
[info]       at zio.ZSchedule.check(ZSchedule.scala:211)
[info]       at zio.ZIO.repeatOrElseEither(ZIO.scala:945)
[info]       at zio.ZIO.as(ZIO.scala:102)
[info]       at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:155)
[info]       at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:146)
[info]       at zio.ZIO.ensuring(ZIO.scala:285)
[info]       at zio.ZIO.onInterrupt(ZIO.scala:568)
[info]       at zio.ZIO.ensuring(ZIO.scala:283)
[info]       at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:146)
[info]       at zio.ZIO$._IdentityFn(ZIO.scala:2392)
[info]       at zio.ZIOFunctions.effectAsyncInterrupt(ZIO.scala:1701)
[info]       at zio.ZIOFunctions.effectAsyncInterrupt(ZIO.scala:1697)
[info]       at zio.ZIOFunctions.effectAsyncInterrupt(ZIO.scala:1697)
[info]       at zio.clock.Clock$Live$$anon$1.sleep(Clock.scala:48)
[info]       at zio.clock.package$.sleep(clock.scala:50)
[info]       at zio.ZQueueSpecUtil$.waitForSize(ZQueueSpec.scala:673)
[info]       at zio.Queue$$anon$1.size(Queue.scala:263)
[info]       at zio.Queue$$anon$1.checkShutdownState(Queue.scala:194)
[info]       at zio.Promise$.poll(Promise.scala:149)
[info]       at zio.Promise$.poll$extension(Promise.scala:149)
[info]       at zio.ZIO.repeatOrElseEither(ZIO.scala:951)
[info]       at zio.ZIO.bracket_(ZIO.scala:131)
[info]       at zio.internal.FiberContext.evaluateNow(FiberContext.scala:469)
[info]       at zio.ZQueueSpec$$anonfun$$lessinit$greater$1.new(ZQueueSpec.scala:54)
[info]       at zio.ZQueueSpec$$anonfun$$lessinit$greater$1.new(ZQueueSpec.scala:52)
[info]       at zio.Ref$.make(Ref.scala:142)
[info]       at zio.ZQueueSpec$$anonfun$$lessinit$greater$1.new(ZQueueSpec.scala:51)
[info]       at zio.ZIO.repeatOrElse(ZIO.scala:930)
[info]       at zio.ZIO.repeatOrElseEither(ZIO.scala:946)
[info]       at zio.ZSchedule.check(ZSchedule.scala:211)
[info]       at zio.ZSchedule.check(ZSchedule.scala:211)
[info]       at zio.ZSchedule$.unfoldM(ZSchedule.scala:746)
[info]       at zio.ZIO.repeatOrElseEither(ZIO.scala:945)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:244)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:252)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:252)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:248)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:244)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:236)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:235)
[info]       at zio.Queue$$anon$1.checkShutdownState(Queue.scala:194)
[info]       at zio.Promise$.poll(Promise.scala:149)
[info]       at zio.Promise$.poll$extension(Promise.scala:149)
[info]       at zio.UIO$.effectSuspendTotal(UIO.scala:147)
[info]       at zio.ZIO.repeatOrElseEither(ZIO.scala:947)
[info]       at <unknown>.<unknown>(ZIO.scala:0)
[info]       at zio.test.mock.MockClock$Mock.sleep(MockClock.scala:64)
[info]       at zio.Promise$.complete$extension(Promise.scala:93)
[info]       at zio.test.mock.MockClock$Mock.sleep(MockClock.scala:58)
[info]       at zio.Ref$.modify$extension(Ref.scala:47)
[info]       at zio.test.mock.MockClock$Mock.sleep(MockClock.scala:57)
[info]       at zio.Promise$.make(Promise.scala:239)
[info]       at zio.clock.package$.sleep(clock.scala:50)
[info]       at zio.ZIO.repeatOrElseEither(ZIO.scala:946)
[info]       at zio.ZSchedule.check(ZSchedule.scala:211)
[info]       at zio.ZSchedule.check(ZSchedule.scala:211)
[info]       at zio.ZSchedule$.unfoldM(ZSchedule.scala:746)
[info]       at zio.ZIO.repeatOrElseEither(ZIO.scala:945)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:244)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:252)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:252)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:248)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:244)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:236)
[info]       at zio.Queue$$anon$1.offerAll(Queue.scala:235)
[info]       at zio.Queue$$anon$1.checkShutdownState(Queue.scala:194)
[info]       at zio.Promise$.poll(Promise.scala:149)
[info]       at zio.Promise$.poll$extension(Promise.scala:149)
[info]       at zio.UIO$.effectSuspendTotal(UIO.scala:147)
[info]       at zio.ZIO.repeatOrElseEither(ZIO.scala:947)
[info]       at <unknown>.<unknown>(ZIO.scala:0)
[info]       at zio.test.mock.MockClock$Mock.sleep(MockClock.scala:64)
[info]       at zio.Promise$.complete$extension(Promise.scala:93)
[info]       at zio.test.mock.MockClock$Mock.sleep(MockClock.scala:58)
[info]       at zio.Ref$.modify$extension(Ref.scala:47)
[info]       at zio.test.mock.MockClock$Mock.sleep(MockClock.scala:57)
[info]       at zio.Promise$.make(Promise.scala:239)
[info]       at zio.clock.package$.sleep(clock.scala:50)
[info]       at zio.ZIO.repeatOrElseEither(ZIO.scala:946)
[info]       at zio.ZSchedule.check(ZSchedule.scala:211)
[info]       at zio.ZSchedule.check(ZSchedule.scala:211)
[info]     
[info]     Fiber:12547 was spawned by:
[info]     
[info]     Fiber:12539 was supposed to continue to:
[info]       a future continuation at zio.ZIO.zipWith(ZIO.scala:1362)
[info]       a future continuation at zio.ZIOFunctions.foreach(ZIO.scala:1832)
[info]       a future continuation at zio.ZIOFunctions.foreachParN(ZIO.scala:1898)
[info]       a future continuation at zio.ZIO.run(ZIO.scala:1007)
[info]       a future continuation at zio.ZIOFunctions.foreachParN(ZIO.scala:1891)
[info]       a future continuation at zio.ZIO.refailWithTrace(ZIO.scala:873)
[info]       a future continuation at zio.test.Spec.foldM(Spec.scala:70)
[info]       a future continuation at zio.ZIO.summarized(ZIO.scala:1115)
[info]       a future continuation at zio.test.TestRunner.run(TestRunner.scala:41)
[info]       a future continuation at zio.ZIO.run(ZIO.scala:1007)
[info]     
[info]     Fiber:12539 execution trace:
[info]       at zio.ZIO.zipWith(ZIO.scala:1362)
[info]       at zio.ZIOFunctions.foreachParN(ZIO.scala:1895)
[info]       at zio.ZIOFunctions.foreachParN(ZIO.scala:1894)
[info]       at zio.ZIOFunctions.foreach(ZIO.scala:1832)
[info]       at zio.ZIOFunctions.foreach(ZIO.scala:1832)
[info]       at zio.ZIOFunctions.foreach(ZIO.scala:1832)
[info]       at zio.ZIOFunctions.foreach(ZIO.scala:1832)
[info]       at zio.ZIOFunctions.foreach(ZIO.scala:1832)
[info]       at zio.ZIOFunctions.foreach(ZIO.scala:1832)
[info]       at zio.ZIOFunctions.foreach(ZIO.scala:1832)
```

**Comment by dkarlinsky** [comment]
I wouldn't mind giving a this a shot. 
I have almost no knowledge of core ZIO - but this could be a great opportunity to dive in :)
Is there code / test case that reproduces this?
(`FiberSpec` doesn't seem to test `interruptAll`)
Mostly as an exercise, I wrote a test for a simple case of `interruptAll`:
```scala
def interruptAll = for {
    semaphore <- Semaphore.make(1)
    _         <- semaphore.acquire
    fibers    <- ZIO.sequence((1 to 10).map{ _ =>
                  semaphore.acquire.fork
                })
    _         <- Fiber.interruptAll(fibers)
    _         <- Fiber.joinAll(fibers)
    is        <- ZIO.sequence(fibers.map(_.poll))
  } yield is.flatten must (forall(beInterrupted) and haveSize(10))

  private def beInterrupted: Matcher[Exit[_, _]] = like[Exit[_,_]] {
    case Exit.Failure(cause) => cause.interrupted must beTrue
  }
```   
As expected it passes (assuming I got the test right).

**Comment by ghostdogpr** [comment]
@dkarlinsky it happened very rarely during one test in ZQueueTest. Here's a build showing the stack trace: https://circleci.com/gh/zio/zio/17876

The test has since been modified to avoid using `interruptChildren` (which uses `interruptAll`), but you can see the old version of the test here: https://github.com/zio/zio/pull/1439/files#diff-588bd7296f8ff886fd4986d54e3de18cL48 (on the left).

I've never seen it locally but if you run that test in a loop, maybe it will happen.

**Comment by ghostdogpr** [comment]
You can have a look at `FiberContext`, where the magic happens (the low-level IO execution loop), in particular `getFibers` which returns the list of fibers that is then interrupted.

**Comment by jdegoes** [comment]
I think this one is caused by `WeakMap` — in particular, by a `FiberContext` being present in the map but with a null value, due to a non-deterministic race condition.

This is just a hunch, but to solve such an issue, after you pull it out of the map, you'd make sure it was non-null, before adding it to the list of returned fibers.

**Comment by dkarlinsky** [comment]
@jdegoes your hunch seems to be on point.
The NPE is triggered in `fs.foldLeft()` [here](https://github.com/zio/zio/blob/062024e3e9abdf92e36fa138a1eafee9eba5433f/core/shared/src/main/scala/zio/Fiber.scala#L296) which is caused by `fs` containing a `null`. 
I traced the null to [this line](https://github.com/zio/zio/blob/a737e76918b59bacb5c5cc072daa989e938e060b/core/shared/src/main/scala/zio/internal/FiberContext.scala#L564), which suggests that the `java.util.Set` in `SuperviseStatus.Supervised` contains a `null`. 
The set is indeed based on a `WeakHashMap`, as seen [here](https://github.com/zio/zio/blob/a737e76918b59bacb5c5cc072daa989e938e060b/core/shared/src/main/scala/zio/internal/FiberContext.scala#L577)

I guess the solution to the NPE is to simply add `.filter(_ != null)` in  [getFibers](https://github.com/zio/zio/blob/a737e76918b59bacb5c5cc072daa989e938e060b/core/shared/src/main/scala/zio/internal/FiberContext.scala#L564). 
Not sure what other consequences supervised fibers being occasionally "forgotten" might have, though ...

**Comment by jdegoes** [comment]
Fixed by #1473
